### ISSUE TITLE
Multiple saved playback states

### ISSUE THREAD

**Issue by OxygenCobalt** [OxygenCobalt/Auxio#407]
Thinking about adding the ability to persist and switch between several playback states at once. 

Normally I might play all my songs on shuffle, but then want to play a specific artist when doing a mentally-intensive task. I'd want to go back to the prior playback state eventually, but #404 would make that impossible as the state would get overriden. 

I might then benefit from something like Musicolet's ability to track several playback states at once. However, not sure how to do this intuitively since this is a bit of an edge-case.

**Comment by KraXen72** [comment]
this is basically just saving a queue to a playlist. good idea but no need to decouple these two ideas imo

**Comment by OxygenCobalt** [comment]
Oh, so that's why that button exists in so many music players. Yeah, my idea is a lot deeper where I want to retain the queue *and* exact position.

**Comment by KraXen72** [comment]
i can see how they are different, but, just for my curiosity, when exactly do you need to save the specific time a song stopped playing?

**Comment by OxygenCobalt** [comment]
It's personal preference, mostly. This is what I mean by that this is a kind of edge-case feature that I'm unsure about implementing.

**Comment by KraXen72** [comment]
oh okay yeah, i get that. honestly by the time i get back to listen to music after a break, unless the song has like >10minutes i'd just play it from the start. but i get why it could be useful for longer songs/spoken word

**Comment by OxygenCobalt** [comment]
Going to go ahead with this. This issue is really annoying, and it would also be a good place to put "Save current queue to playlist" functionality.

**Comment by ildar** [comment]
this discussion reminds me what I wished a long time ago: Playlists with positions. So that if I switch *from* listening of a playlist, the app saves this position (of that playlist) and when I'm back to it, it goes from where I stopped last time

**Comment by etyarews** [comment]
What would be the UX plan for this? I do quite like the idea, but I fail to see a proper intuitive way to use this without bringing some issues with other features

**Comment by OxygenCobalt** [comment]
My idea @etyarews is roughly:
- A new button to open the playback state is added alongside the equalizer button.
- The button will open a dialog containing a list of all current playback states.
- If you click on a playback state, Auxio will immediately switch to it.
- Each playback state item would also have a delete button, and there would be an additional "New" item that creates a playback state from the current state.
- There would probably be some pruning system for "dead" playback states to avoid Auxio consuming excessive amounts of space, but I'm not sure how that would exactly play out.

**Comment by etyarews** [comment]
Hmm, my first thought would be to merge it with Playlists as there is a bit of overlap between the two (i.e. list of songs).

I'd advise for Auxio to keep track of how many songs were played for a given playback state. If only one song was played, then it can be basically nuked without care.

**Comment by OxygenCobalt** [comment]
> Hmm, my first thought would be to merge it with Playlists as there is a bit of overlap between the two (i.e. list of songs).

There's not really that much overlap. Yes, there's the queue list, but there is also position, repeat mode, shuffle state, etc. that is not encoded in playlists.

> I'd advise for Auxio to keep track of how many songs were played for a given playback state. If only one song was played, then it can be basically nuked without care.

That's a good idea. Might not fully solve the problem though.

**Comment by etyarews** [comment]
Hmm, just to clarify: I mean that playback stats would be inside the Playlist tab, but would be under a sub category, like a subheader divider.

Playstates could also be a vertical list of cards or something different, like "continue where you left off" type of deal.

**Comment by OxygenCobalt** [comment]
> Hmm, just to clarify: I mean that playback stats would be inside the Playlist tab, but would be under a sub category, like a subheader divider.

Yeah, that still feels very disjoint. In addition, this feature is arguably a little power user-y and I don't want to force it onto users.

> Playstates could also be a vertical list of cards or something different, like "continue where you left off" type of deal.

Eh, that veers into "For You" page territory, which I personally don't like.

**Comment by KraXen72** [comment]
i think you should save this data for playlists, but don't do anything with it by default (and keep only the latest).
  
Then, you'd have a toggle for power-users under the 'save playback state' button in settings (if they are there they probably want that), which would enable a new tab like playlists, but for playback states.
When they enable it, it would suggest creating a playback state from one of their recently listened playlists / from any playlists that they've played in the past & has the data.

When turned off again, the most recent playback states (applicable) would get applied to playlists, and 'hidden' again.

**Comment by OxygenCobalt** [comment]
I don't fully understand, but I actually like the idea of having an enable toggle. It avoids useless playback states from piling up unless someone is already aware of the feature.

**Comment by KraXen72** [comment]
- if the toggle is off: auxio only keeps 1 playback state internally for each playlist. nothing else (custom queues)
- if user turns on toggle: auxio suggests to "import" these internal playback states into a new tab with playback states
- if toggle is on: user can save (from queue view) & resume playback states (up to x recent (autosaved) / manually saved)
- if user turns off toggle: auxio looks at saved & recent playbacks states, and assignes each playlist a playback state internally if applicable, so the cycle can continue
  
this is just an idea so far, but i think it could work

**Comment by etyarews** [comment]
Basically, Playlists are semi-fused with Playback. Meaning that each playlist also has a playback state alongside it, and users selecting a playlist can "continue from where they've left off", or they can start the playlist again from the beginning.

That is, IMO, a better solution. Randomly playing songs doesn't save a playback state on this mode, but playing from a playlist does. A toggle in the settings would expose the current list of playback states saved and uncouple it from the playlists, allowing *more* playback states, but still allowing basic users to have something similar.

**Comment by OxygenCobalt** [comment]
Hm. A little interesting, but I don't know how that could be integrated into the UI @KraXen72 @etyarews.

**Comment by etyarews** [comment]
I don't think there is much to think about integrating into the UI. When the user toggles "advanced playback states" or whatever it is going to be called, a new tab on the home screen appears and that is that. When it doesn't, any time they play something that could continue from an existing playback state, a bottom menu appears asking if they want to pick up they left off or if they want to start from the beginning.

Basically, any interaction other than playing a song directly on the "Songs" tab should create a playback state automatically. And then you can remove unused playback states with a couple of rules, like for instance:

Automatic Playback states are automatically removed if there are more than 10 states saved AND they are older than 2 weeks.

**Comment by OxygenCobalt** [comment]
Thanks for the extensive feature description. I'll keep this in mind when working on this.

**Comment by brighteyed** [comment]
Just discovered Auxio while searching for an offline player for podcasts and audiobooks and I'm impressed with its very smooth UI/UX. 
For my use case I could add all podcast's episodes into a playlist and toggle something like `Autosave state` mode for this playlist. This would allow me to switch between such "advanced" playlists and resume playback where I left off.

**Comment by KraXen72** [comment]
@brighteyed maybe unrelated, but you might wanna check out the open source [AntennaPod](https://github.com/AntennaPod/AntennaPod) podcast player.

**Comment by brighteyed** [comment]
> @brighteyed maybe unrelated, but you might wanna check out the open source [AntennaPod](https://github.com/AntennaPod/AntennaPod) podcast player.

Thanks for the suggestion! I've found AntennaPod quite feature rich for podcasts. Will try to use it for audiobooks as well since it supports local folders.

**Comment by OxygenCobalt** [comment]
I also agree with @KraXen72 @brighteyed. Generally I attempt to optimize features for music and not podcasts or audiobooks. Not that you can't use Auxio with such, it's just always going to be sub-par compared to an app designed for podcasts/audiobooks like AntennaPod.
